### ISSUE TITLE
JSONRPCClientReplyError: replacement transaction underpriced

### ISSUE THREAD

**Issue by LefterisJP** [raiden-network/raiden#804]
## Problem Definition

We have been getting `JSONRPCClientReplyError: replacement transaction underpriced`.

This happens in multiple tests. It has happened in:

- [test_new_netting_contract](https://gist.github.com/LefterisJP/590866cd904df439daf9d364f69eb39b)
- [test_settlement](https://gist.github.com/LefterisJP/99cde3078ee517e62fd4aba4343119bb)
- others

## Discussion and Path to Eventual Solution

We are still investigating why this happens. Geth has [two](github.com/ethereum/go-ethereum/blob/master/core/tx_pool.go#L51) [places](github.com/ethereum/go-ethereum/blob/master/core/tx_pool.go#L463) in the tx_pool code where they are using that error.

It's emitted only when another transaction with the same nonce as the previous transaction is emitted and the `gas price bump` is not met.

Geth has the possibility to define the gas price bump as a command line argument `--txpool.pricebump` so we are also playing with that possibility [here](https://github.com/raiden-network/raiden/pull/787/files)

Ofcourse the question should also be why do we try to use the same nonce for a pending transaction in order to replace it in the first place??

## Tasklist

- Investigate, find why this happens and fix it

**Comment by LefterisJP** [comment]
This also happened while testing in the wild with geth 1.6.6 in ropsten.

I made two requests to the API to create two different channels sumultaneously but with 2 different partners.

The first failed:

```
http PUT http://localhost:5001/api/1/channels partner_address="0x86e610c0645c58f352996c32ed9d11e77558ff5b" token_address="0x0f114a1e9db192502e7856309cc899952b3db1ed" balance=40 reveal_timeout=30 settle_timeout=65 --timeout=99999
HTTP/1.1 500 INTERNAL SERVER ERROR                                  
Content-Length: 37                                                  
Content-Type: application/json                                      
Date: Wed, 23 Aug 2017 09:15:12 GMT                                 

{                                                                   
    "message": "Internal Server Error"                              
}                                             
```

and the second succeeded

```
http PUT http://localhost:5001/api/1/channels partner_address="0x25511699c252eea2678266857c98f459df97b77c" token_address="0x0f114a1e9db192502e7856309cc899952b3db1ed" balance=40 reveal_timeout=30 settle_timeout=65 --timeout=99999
HTTP/1.1 201 CREATED                                                
Content-Length: 249                                                 
Content-Type: application/json                                      
Date: Wed, 23 Aug 2017 09:15:51 GMT                                 
mimetype: application/json                                          

{                                                                   
    "balance": 40,                                                  
    "channel_address": "0x27773fa66ecf0e59dee0baa9a2a34225a49a4e7c",
    "partner_address": "0x25511699c252eea2678266857c98f459df97b77c",
    "settle_timeout": 65,                                           
    "state": "opened",                                              
    "token_address": "0x0f114a1e9db192502e7856309cc899952b3db1ed"   
}                                                 
```

In the raiden logs we got the replacement transaction underpriced error for the deposit part of the opening:

```
INFO:raiden.network.protocol    MESSAGE RECEIVED node:9bed7fd1 echohash:1b47f73e <Ping [msghash=0eed6060]>                              
INFO:raiden.message_handler     message received message=<Ping [msghash=0eed6060]>                                                      
INFO:raiden.network.protocol    SENDING ACK 9bed7fd1 > 9598decb <Ack [echohash:1b47f73e]>                                               
INFO:raiden.network.protocol    MESSAGE RECEIVED node:9bed7fd1 echohash:41cb0eef <Ping [msghash=6eceb03e]>                              
INFO:raiden.message_handler     message received message=<Ping [msghash=6eceb03e]>                                                      
INFO:raiden.network.protocol    SENDING ACK 9bed7fd1 > ba880131 <Ack [echohash:41cb0eef]>                                               
[2017-08-23 11:15:12,111] ERROR in app: Exception on /api/1/channels [PUT]                                                              
Traceback (most recent call last):                                                                                                      
  File "/home/lefteris/.virtualenvs/raiden/lib/python2.7/site-packages/flask/app.py", line 1612, in full_dispatch_request               
    rv = self.dispatch_request()                                                                                                        
  File "/home/lefteris/.virtualenvs/raiden/lib/python2.7/site-packages/flask/app.py", line 1598, in dispatch_request                    
    return self.view_functions[rule.endpoint](**req.view_args)                                                                          
  File "/home/lefteris/.virtualenvs/raiden/lib/python2.7/site-packages/flask_restful/__init__.py", line 480, in wrapper                 
    resp = resource(*args, **kwargs)                                                                                                    
  File "/home/lefteris/.virtualenvs/raiden/lib/python2.7/site-packages/flask/views.py", line 84, in view                                
    return self.dispatch_request(*args, **kwargs)                                                                                       
  File "/home/lefteris/.virtualenvs/raiden/lib/python2.7/site-packages/flask_restful/__init__.py", line 595, in dispatch_request        
    resp = meth(*args, **kwargs)                                                                                                        
  File "/home/lefteris/.virtualenvs/raiden/lib/python2.7/site-packages/webargs/core.py", line 441, in wrapper                           
    return func(*args, **kwargs)                                                                                                        
  File "/home/lefteris/ew/raiden/raiden/api/v1/resources.py", line 54, in put                                                           
    return self.rest_api.open(**kwargs)                                                                                                 
  File "/home/lefteris/ew/raiden/raiden/api/rest.py", line 275, in open                                                                 
    balance                                                                                                                             
  File "/home/lefteris/ew/raiden/raiden/api/python.py", line 220, in deposit                                                            
    netting_channel.deposit(amount)                                                                                                     
  File "/home/lefteris/ew/raiden/raiden/network/rpc/client.py", line 1026, in deposit                                                   
    amount,                                                                                                                             
  File "/home/lefteris/ew/raiden/raiden/network/rpc/client.py", line 274, in estimate_and_transact                                      
    gasprice=classobject.gasprice                                                                                                       
  File "/home/lefteris/.virtualenvs/raiden/lib/python2.7/site-packages/pyethapp/rpc_client.py", line 783, in transact                   
    **kargs                                                                                                                             
  File "/home/lefteris/ew/raiden/raiden/network/rpc/client.py", line 148, in send_transaction                                           
    data_encoder(rlp.encode(tx)),                                                                                                       
  File "/home/lefteris/.virtualenvs/raiden/lib/python2.7/site-packages/pyethapp/rpc_client.py", line 413, in call                       
    raise JSONRPCClientReplyError(jsonrpc_reply.error)                                                                                  
JSONRPCClientReplyError: replacement transaction underpriced                                                                            
127.0.0.1 - - [2017-08-23 11:15:12] "PUT /api/1/channels HTTP/1.1" 500 164 94.504670                                                    
INFO:raiden.event_handler       state_change received state_change=<raiden.transfer.mediated_transfer.state_change.ContractReceiveTokenAdded object at 0x7f869e96dc10>
INFO:raiden.event_handler       state_change received state_change=<raiden.transfer.mediated_transfer.state_change.ContractReceiveNewChannel object at 0x7f869e9224d0>
```

So the channel was opened succesfully but the deposit transaction failed.

**Comment by hackaugusto** [comment]
I believe this is fixed, has anyone experience problems with this?

**Comment by LefterisJP** [comment]
@hackaugusto Any idea where this was fixed? Via https://github.com/raiden-network/raiden/pull/974? I think the reason it was not closed backthen is that the issue kept persisting after that.

**Comment by andrevmatos** [comment]
[I've a branch](https://github.com/andrevmatos/raiden/tree/parallel_join) that have a code that definitely triggered the issue in the past (parallel join, [open+deposit] multiple channels in parallel on join/connect to network, instead of waiting for previous channel for open+deposit to be confirmed before proceeding). But I'm not able to test it properly, I think I've some nat issue in my network. Can someone check the branch, to see if Join Network works? If so, this issue is fixed.

**Comment by hackaugusto** [comment]
> Any idea where this was fixed? Via #974?

That is my guess.

> I think the reason it was not closed backthen is that the issue kept persisting after that.

I'm aware this can and does happen on production https://github.com/raiden-network/raiden/issues/1061 , but I personally haven't seen this error on our test suite anymore.

**Comment by LefterisJP** [comment]
As discussed in the Sprint retrospective meeting this issue is considered closed as it has not been reproduced again.

If anyone manages to reproduce it, then please make as detailed a report as possible about it and reopen this issue.
